<div style="width: 100%; clear: both;">
<div style="float: left; width: 50%;">
<img src="https://www.uoc.edu/portal/_resources/common/imatges/sala_de_premsa/noticies/2016/202-nova-marca-uoc.jpg", align="left" width="380" height="120">
</div>
</div>
<div style="float: right; width: 50%;">
<p style="margin: 0; padding-top: 22px; text-align:right;">Visualització de Dades</p>
<p style="margin: 0; text-align:right;">Màster en Ciència de Dades</p>
<p style="margin: 0; text-align:right; padding-button: 100px;">Estudis d'Informàtica, Multimèdia i Telecomunicacions</p>
</div>
</div>
<div style="width: 100%; clear: both;">
<div style="width:100%;">&nbsp;</div>

Arnau Gusart Verdú

# PRA Part 1

En la primera part de la pràctica seleccionarem un dataset i comentarem els següents aspectes:

<ul>
<li>1. Justificació de la selecció.
<li>2. Rellevancia del conjunt de dades.
<li>3. Complexitat del conjunt de dades.
<li>4. Originalitat del conjunt de dades.
<li>5. Preguntes a respondre amb el conjunt de dades.
</ul>
  

### 1.- Justificació i rellevància de la selecció

Dataset triat:
Qualitat de l'aire als punts de mesurament automàtics de la xarxa de vigilància i previsió de la contaminació atmosfèrica.
Enllaç: https://datos.gob.es/es/catalogo/a09002970-calidad-del-aire-en-los-puntos-de-medida-automaticos-de-la-red-de-vigilancia-y-prevision-de-la-contaminacio-atmosferica

L'elecció d'aquest dataset ha estat per motius personals. L'aire al igual que l'aigua és un bé essencial que donem per sentat. Des de fa uns anys (concretament des de la revolució industrial) que hem massificat processos de combustió i altres processos químics que desprenen components adicionals al aire. A moltes ciutats amb molta població com ara Barcelona és facil veure els efectes de la contaminació de l'aire sobre tot quan no plou. Crec que fer visualitzacions amb aquest dataset ens pot permetre reflexionar sobre les condicions actuals que tenim a Catalunya i donar-nos conciència.

### 2.- Complexitat del dataset

Carreguem el dataset per fer un anàlisi ràpid:

In [3]:
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
import warnings
from google.colab import drive
import os

warnings.filterwarnings('ignore')
drive.mount('/content/drive')
os.listdir('/content/drive/My Drive/')
os.chdir('/content/drive/My Drive/')

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


In [4]:
# Carreguem el fitxer de ressenyes:
df = pd.read_csv('Qualitat_de_l_aire_als_punts_de_mesurament_autom_tics_de_la_Xarxa_de_Vigil_ncia_i_Previsi__de_la_Contaminaci__Atmosf_rica.csv')

df = pd.DataFrame(df)
print('Mida del dataset', df.shape)

df.sample(5, random_state=42)

Mida del dataset (3284257, 40)


CODI EOI                               NOM ESTACIO        DATA  \
857171   43148003                     Tarragona (Bonavista)  05/05/2017   
1642014   8089004               Gavà (c/Girona - c/Progrés)  27/11/2000   
1863066   8073001  Cornellà de Llobregat (Allende - Bonveí)  23/10/2008   
793012    8169008  El Prat de Llobregat (Jardins de la Pau)  06/04/2015   
307126   43005002                                   Alcover  24/08/2016   

         MAGNITUD CONTAMINANT UNITATS TIPUS ESTACIO AREA URBANA  CODI INE  \
857171         65         H2S   µg/m3    industrial  peri-urban     43148   
1642014        10        PM10   µg/m3       traffic  peri-urban      8089   
1863066         7          NO   µg/m3       traffic       urban      8073   
793012          1         SO2   µg/m3    background  peri-urban      8169   
307126         65         H2S   µg/m3    industrial       rural     43005   

                      MUNICIPI  ...   19h   20h   21h   22h   23h    24h  \
857171               Tarragona  ...   1.0   1.0   2.3   1.4   1.1    1.0   
1642014                   Gavà  ...   1.0   1.0   1.0  26.0  71.0  123.0   
1863066  Cornellà de Llobregat  ...  27.0  73.0  79.0  67.0  55.0   32.0   
793012   Prat de Llobregat, el  ...   3.0   3.0   3.0   2.0   2.0    2.0   
307126                 Alcover  ...   1.0   1.0   1.0   1.0   1.0    1.0   

         ALTITUD    LATITUD  LONGITUD               Georeferència  
857171        39  41.115910  1.191999   POINT (1.191999 41.11591)  
1642014       12  41.300574  2.013889  POINT (2.013889 41.300574)  
1863066       27  41.356380  2.076113   POINT (2.076113 41.35638)  
793012         5  41.321487  2.097702  POINT (2.097702 41.321487)  
307126       238  41.278687  1.179898  POINT (1.179898 41.278687)  

[5 rows x 40 columns]

Tenim un dataset doncs amb més de 3 milions de registres presos als punts de mesurament automàtics de la Xarxa de Vigilància i Previsió de la Contaminació Atmosfèrica des de l’any 1991 fins al dia anterior a l’actual per tot Catalunya amb 40 variables diferents.

In [ ]:
df.dtypes

CODI EOI           int64
NOM ESTACIO       object
DATA              object
MAGNITUD           int64
CONTAMINANT       object
UNITATS           object
TIPUS ESTACIO     object
AREA URBANA       object
CODI INE           int64
MUNICIPI          object
CODI COMARCA       int64
NOM COMARCA       object
01h              float64
02h              float64
03h              float64
04h              float64
05h              float64
06h              float64
07h              float64
08h              float64
09h              float64
10h              float64
11h              float64
12h              float64
13h              float64
14h              float64
15h              float64
16h              float64
17h              float64
18h              float64
19h              float64
20h              float64
21h              float64
22h              float64
23h              float64
24h              float64
ALTITUD            int64
LATITUD          float64
LONGITUD         float64
Georeferència     object


Tenim els tipus de variables següents:

- Categòriques:
Nom estació, Contaminant, Unitats, Tipus estació, Àrea urbana, Municipi, Codi i Nom comarca.
- Numèriques:
Codi EOI, data, Magnitud, Codi INE, Mesures per hora (01h, 02h ...).
- Geospacials:
Altitud (també la podriem contar com numèrica), Latitud, Longitud, Georeferència.

A més a més podrem combinar variables per obtenir promitjos per dia/mes/any o promitjos per municipi/comarca per poder extreure dades més completes.

Aparentment tenim un dataset bastant complet amb dades de diferents tipus.


### 3.- Questions a Respondre

La qualitat de l'aire es un tema que si que es tracta i s'estudia, per exemple quan donen el temps al telenotícies o apareix fins i tot en els telèfons i rellotges intel·ligents. Però crec que amb el conjunt de mesures dels últims 30 anys podem extreure informació a més gran escala, comparant l'evolució de diferents zones. Un bon complement seria agafar un dataset de la generalitat que evalui el nivell d'industrialització de les ciutats per comparar amb els nivells de contaminació per exemple. També podem aprofitar per visualitzar l'efecte que va tenir la Covid19 a la qualitat del aire.



### 4.- Diccionari de variables

In [ ]:
# Recompte de nuls
null_sum = df.isnull().sum()
null_percentage = (null_sum / len(df)) * 100
null_percentage = null_percentage.round(2)
null_info = pd.concat([null_sum, null_percentage], axis=1)
null_info.columns = ['Recompte de Nuls', '% Nuls']

print(null_info)

               Recompte de Nuls  % Nuls
CODI EOI                      0    0.00
NOM ESTACIO                   0    0.00
DATA                          0    0.00
MAGNITUD                      0    0.00
CONTAMINANT                   0    0.00
UNITATS                       0    0.00
TIPUS ESTACIO                 0    0.00
AREA URBANA                   0    0.00
CODI INE                      0    0.00
MUNICIPI                      0    0.00
CODI COMARCA                  0    0.00
NOM COMARCA                   0    0.00
01h                      114030    3.47
02h                      121660    3.70
03h                      106689    3.25
04h                      101877    3.10
05h                       99046    3.02
06h                       98037    2.99
07h                       98218    2.99
08h                      100713    3.07
09h                      111272    3.39
10h                      124840    3.80
11h                      150165    4.57
12h                      168804    5.14


Veiem que com a molt tenim un 5% de nuls en les variables de mesura de la qualitat de l'aire i en la georeferència, que no hauria de ser pas massa problema, podem triar ignorar aquests registres ja que en tenim molts o en el cas de la georeferència podem optar per fer servir les altres variables geospacials.

#### Variables numèriques

In [ ]:
df['CODI EOI'].describe()

count    3.284257e+06
mean     1.826647e+07
std      1.499344e+07
min      8.015001e+06
25%      8.113007e+06
50%      8.245012e+06
75%      4.300400e+07
max      4.317100e+07
Name: CODI EOI, dtype: float64

El codi EOI pren valors de 8 milions a 43 milions, sembla que més de la meitat de codis voren el valor de 8 milions.

In [ ]:
df['DATA'].describe()

count        3284257
unique         12180
top       06/03/2023
freq             382
Name: DATA, dtype: object

En el cas de la data tenim més de 12180 dates diferents des del 1991.

In [ ]:
df['MAGNITUD'].describe()

count    3.284257e+06
mean     1.246570e+01
std      1.799781e+01
min      1.000000e+00
25%      7.000000e+00
50%      8.000000e+00
75%      1.200000e+01
max      3.310000e+02
Name: MAGNITUD, dtype: float64

Els valors de magnitud van des de 1 fins a 300 aproximadament tot i que més del 75% és menor, és possible que hi hagi aquests valors outliers ja que el CO té unes unitats de mesura diferents.

In [ ]:
df['CODI INE'].describe()

count    3.284257e+06
mean     1.826646e+04
std      1.499344e+04
min      8.015000e+03
25%      8.113000e+03
50%      8.245000e+03
75%      4.300400e+04
max      4.317100e+04
Name: CODI INE, dtype: float64

En aquest cas tenim un nou codi identificador que aquest cop va des de 8000 fins a 43000.

In [ ]:
df[['01h','02h','03h','04h','05h','06h','07h','08h','09h','10h','11h','12h']].describe()

01h           02h           03h           04h           05h  \
count  3.170227e+06  3.162597e+06  3.177568e+06  3.182380e+06  3.185211e+06   
mean   2.094909e+01  1.941331e+01  1.809138e+01  1.719751e+01  1.678046e+01   
std    3.110089e+01  2.876023e+01  2.693434e+01  2.573668e+01  2.510392e+01   
min   -2.800000e+00 -2.800000e+00 -2.800000e+00 -2.800000e+00 -2.800000e+00   
25%    1.000000e+00  1.000000e+00  1.000000e+00  1.000000e+00  1.000000e+00   
50%    8.000000e+00  8.000000e+00  7.000000e+00  7.000000e+00  6.600000e+00   
75%    3.000000e+01  2.800000e+01  2.500000e+01  2.400000e+01  2.300000e+01   
max    1.608000e+03  1.582000e+03  1.652000e+03  2.222000e+03  2.222000e+03   

                06h           07h           08h           09h           10h  \
count  3.186220e+06  3.186039e+06  3.183544e+06  3.172985e+06  3.159417e+06   
mean   1.790893e+01  2.112288e+01  2.704606e+01  3.139185e+01  3.030948e+01   
std    2.655440e+01  3.266452e+01  4.497472e+01  5.353119e+01  4.938315e+01   
min   -2.850000e+00 -2.850000e+00 -2.850000e+00 -2.800000e+00 -2.800000e+00   
25%    1.000000e+00  1.000000e+00  1.900000e+00  2.000000e+00  2.000000e+00   
50%    8.000000e+00  9.000000e+00  1.100000e+01  1.200000e+01  1.300000e+01   
75%    2.500000e+01  2.900000e+01  3.600000e+01  4.100000e+01  4.100000e+01   
max    2.222000e+03  2.517000e+03  2.304000e+03  8.777000e+03  7.777000e+03   

                11h           12h  
count  3.134092e+06  3.115453e+06  
mean   2.743201e+01  2.576133e+01  
std    4.134341e+01  3.698865e+01  
min   -2.850000e+00 -2.850000e+00  
25%    2.000000e+00  2.000000e+00  
50%    1.200000e+01  1.100000e+01  
75%    3.900000e+01  3.700000e+01  
max    7.777000e+03  7.777000e+03

In [ ]:
df[['13h','14h','15h','16h','17h','18h','19h','20h','21h','22h','23h','24h']].describe()

13h           14h           15h           16h           17h  \
count  3.117057e+06  3.139159e+06  3.157003e+06  3.167470e+06  3.184651e+06   
mean   2.468904e+01  2.404176e+01  2.326284e+01  2.301201e+01  2.330448e+01   
std    3.498906e+01  3.431185e+01  3.397918e+01  3.405611e+01  3.429838e+01   
min   -2.800000e+00 -2.800000e+00 -2.800000e+00 -2.800000e+00 -2.800000e+00   
25%    2.000000e+00  2.000000e+00  1.300000e+00  1.000000e+00  1.000000e+00   
50%    1.000000e+01  1.000000e+01  8.000000e+00  8.000000e+00  8.000000e+00   
75%    3.500000e+01  3.300000e+01  3.100000e+01  3.000000e+01  3.100000e+01   
max    7.777000e+03  7.777000e+03  7.777000e+03  7.777000e+03  7.777000e+03   

                18h           19h           20h           21h           22h  \
count  3.194053e+06  3.197303e+06  3.198055e+06  3.197443e+06  3.196867e+06   
mean   2.464025e+01  2.621049e+01  2.734589e+01  2.759533e+01  2.684521e+01   
std    3.597921e+01  3.897643e+01  4.127764e+01  4.185641e+01  4.070729e+01   
min   -2.800000e+00 -2.800000e+00 -2.800000e+00 -2.850000e+00 -2.800000e+00   
25%    1.100000e+00  1.200000e+00  1.300000e+00  1.300000e+00  1.200000e+00   
50%    9.000000e+00  9.000000e+00  1.000000e+01  1.000000e+01  1.000000e+01   
75%    3.400000e+01  3.700000e+01  4.000000e+01  4.000000e+01  3.900000e+01   
max    1.299000e+03  2.116000e+03  1.650000e+03  1.405000e+03  2.052000e+03   

                23h           24h  
count  3.193907e+06  3.180101e+06  
mean   2.512456e+01  2.275144e+01  
std    3.791263e+01  3.403457e+01  
min   -2.800000e+00 -2.800000e+00  
25%    1.000000e+00  1.000000e+00  
50%    1.000000e+01  9.000000e+00  
75%    3.700000e+01  3.300000e+01  
max    1.853000e+03  1.511000e+03

Per les mesures per hora tenim mínims molt similars, però si evaluem els màxims veiem que durant les hores de laborables es pren els valors més elevats.
Les hores on tenim més mesures nules son entre les 12h i 13h.

In [ ]:
df['ALTITUD'].describe()

count    3.284257e+06
mean     1.719265e+02
std      2.612567e+02
min      3.000000e+00
25%      3.300000e+01
50%      7.800000e+01
75%      1.920000e+02
max      1.570000e+03
Name: ALTITUD, dtype: float64

Veiem que la majoria de mesures estan preses vora del nivell del mar (200m com a molt) Tot i que tenim mesures que arriben als 1570m d'alçada.

#### Variables Categòriques

In [ ]:
df['NOM ESTACIO'].describe()

count      3284257
unique         116
top       Igualada
freq         79857
Name: NOM ESTACIO, dtype: object

Tenim 116 estacions de mesura diferents sent la de Igualada la que té més mesures.

In [ ]:
print(df['CONTAMINANT'].describe())
print(df['CONTAMINANT'].unique())

count     3284257
unique         17
top           NO2
freq       577623
Name: CONTAMINANT, dtype: object
['NOX' 'O3' 'NO' 'NO2' 'PM10' 'CO' 'SO2' 'PM2.5' 'C6H6' 'H2S' 'PM1' 'Hg'
 'PS' 'HCT' 'Cl2' 'HCNM' 'HCl']


Tenim mesures de 17 contaminants diferents de les quals NO2 és la categoria amb més mesures.

In [ ]:
print(df['UNITATS'].describe())
print(df['UNITATS'].unique())

count     3284257
unique          5
top         µg/m3
freq      2884943
Name: UNITATS, dtype: object
['µg/m3' 'mg/m3' 'ng/m3' 'ug/m3' 'ppm']


La majoria d'unitats és mesuren en micrograms per metre cúbic, pero tenim 4 unitats diferents per els contaminants que en són l'excepció.

In [ ]:
print(df['TIPUS ESTACIO'].describe())
print(df['TIPUS ESTACIO'].unique())

count        3284257
unique             3
top       background
freq         1488807
Name: TIPUS ESTACIO, dtype: object
['traffic' 'background' 'industrial']


Tenim 3 tipus d'estació de mesura, la meitat de les quals son de tipus background.

In [ ]:
print(df['AREA URBANA'].describe())
print(df['AREA URBANA'].unique())

count     3284257
unique          4
top         urban
freq      1253484
Name: AREA URBANA, dtype: object
['urban' 'suburban' 'rural' 'peri-urban']


Tenim 4 tipus de arees urbanes segons la proximitat que tinguin a les ciutats. Un terç de les mesures estan preses en àrea urbana.

In [ ]:
print(df['CODI COMARCA'].describe())
print(df['CODI COMARCA'].unique())
print(df['NOM COMARCA'].describe())
print(df['NOM COMARCA'].unique())

count    3.284257e+06
mean     2.274982e+01
std      1.305532e+01
min      1.000000e+00
25%      1.100000e+01
50%      1.800000e+01
75%      3.600000e+01
max      4.100000e+01
Name: CODI COMARCA, dtype: float64
[41 13  6 11 24 40 21 29 14 17  8  7 36 15  3 20 18 25 26  9  2 10 37 31
 19 33 23 22  1 30]
count        3284257
unique            31
top       Barcelonès
freq          546312
Name: NOM COMARCA, dtype: object
['Vallès Oriental' 'Barcelonès' 'Anoia' 'Baix Llobregat' 'Osona'
 'Vallès Occidental' 'Maresme' 'Priorat' 'Berguedà' 'Garraf' 'Baix Camp'
 'Bages' 'Tarragonès' 'Cerdanya' 'Alt Penedès' 'Gironès' 'Garrigues'
 'Pallars Jussà' 'Pallars Sobirà' 'Baix Ebre' 'Alt Empordà' 'Baix Empordà'
 'Terra Alta' 'Ripollès' 'Garrotxa' 'Segrià' 'Noguera' 'Montsià'
 'Alt Camp' 'Ribera d¿Ebre' "Ribera d'Ebre"]


En aquest cas el codi fa referència a la categoria (Nom de comarca) per tant totes dues ens donen la mateixa informació de dues maneres diferents. Veiem que la comarca amb més mesures es el Barcelonès.

#### Variables Geospacials

In [ ]:
latitude_stats = df['LATITUD'].describe()
longitude_stats = df['LONGITUD'].describe()


print(latitude_stats)
print("\n")
print(longitude_stats)

count    3.284257e+06
mean     4.145879e+01
std      3.442169e-01
min      4.055281e+01
25%      4.121899e+01
50%      4.142559e+01
75%      4.157840e+01
max      4.526166e+01
Name: LATITUD, dtype: float64


count    3.284257e+06
mean     1.795685e+00
std      5.597477e-01
min      2.797220e-01
25%      1.236701e+00
50%      2.007398e+00
75%      2.188298e+00
max      3.212901e+00
Name: LONGITUD, dtype: float64


Aquestes variables prenen els valors dins de Catalunya. Latitud: 40º-45ºN Longitud: 0,5º-3ºE

In [ ]:
print(df['Georeferència'].describe())

count                        3123607
unique                           201
top       POINT (1.151943 41.111956)
freq                           63476
Name: Georeferència, dtype: object


Manera alternativa de representar el parell de valors Longitud-Latitud.

In [14]:
df['DATA'] = pd.to_datetime(df['DATA'])
df['Year'] = df['DATA'].dt.year

# Reorganitzem les columnes
df_melted = df.sample(100000, random_state=42).melt(
    id_vars=['NOM ESTACIO', 'DATA', 'Year', 'MAGNITUD', 'CONTAMINANT', 'UNITATS', 'TIPUS ESTACIO', 'AREA URBANA', 'CODI INE', 'MUNICIPI', 'CODI COMARCA', 'NOM COMARCA', 'LATITUD', 'LONGITUD'],
    value_vars=[f'{i:02d}h' for i in range(1, 25)],
    var_name='Hour',
    value_name='Value'
)

df_melted['Hour'] = df_melted['Hour'].str.slice(0, 2).astype(int)

# Promig per hores de cada any
hourly_avg_df = df_melted.groupby(['Year', 'MUNICIPI', 'Hour', 'CONTAMINANT'])['Value'].mean().reset_index()

# Promig anual per capitals de provincia
main_cities = ['Barcelona', 'Tarragona', 'Girona', 'Lleida']
yearly_avg_df = df_melted[df_melted['MUNICIPI'].isin(main_cities)].groupby(['Year', 'MUNICIPI', 'CONTAMINANT'])['Value'].mean().reset_index()

print("Hourly Average DataFrame:")
print(hourly_avg_df.head())

print("\nYearly Average DataFrame for Main Cities:")
print(yearly_avg_df.head())


Hourly Average DataFrame:
   Year  MUNICIPI  Hour CONTAMINANT      Value
0  1991  Badalona     1          CO   0.841667
1  1991  Badalona     1          NO  40.416667
2  1991  Badalona     1         NO2  38.750000
3  1991  Badalona     1          O3  19.125000
4  1991  Badalona     1          PS  73.285714

Yearly Average DataFrame for Main Cities:
   Year   MUNICIPI CONTAMINANT      Value
0  1991  Barcelona          CO   1.416087
1  1991  Barcelona        HCNM   1.065263
2  1991  Barcelona         HCT   1.495000
3  1991  Barcelona          NO  58.527500
4  1991  Barcelona         NO2  63.848411


In [15]:
import plotly.express as px
year = 2024
filtered_df = hourly_avg_df[hourly_avg_df['Year'] == year]

fig = px.line(filtered_df, x='Hour', y='Value', color='MUNICIPI',
              title=f'Hourly Average Concentrations for Each City in {year}',
              labels={'Hour': 'Hour of the Day', 'Value': 'Average Concentration', 'MUNICIPI': 'City'})
fig.update_layout(legend_title_text='City')
fig.show()


In [9]:
fig = px.bar(yearly_avg_df, x='Year', y='Value', color='MUNICIPI', barmode='group',
             title='Yearly Average Concentrations for Main Cities',
             labels={'Year': 'Year', 'Value': 'Average Concentration', 'MUNICIPI': 'City'})
fig.update_layout(legend_title_text='City')
fig.show()


In [11]:
import plotly.graph_objects as go

filtered_df = hourly_avg_df[hourly_avg_df['MUNICIPI'] == city]
hours = filtered_df['Hour'].unique()
years = filtered_df['Year'].unique()
z_data = filtered_df.pivot(index='Hour', columns='Year', values='Value').values

fig = go.Figure(data=[go.Surface(z=z_data, x=years, y=hours)])

fig.update_layout(title=f'3D Surface Plot of Hourly Concentrations Over the Year for {city}',
                  scene=dict(xaxis_title='Year', yaxis_title='Hour of the Day', zaxis_title='Average Concentration'))
fig.show()


In [60]:
import plotly.express as px

# Filtrem per any i contaminant
year = 2024
contaminant = 'NO2'
filtered_df = hourly_avg_df[(hourly_avg_df['Year'] == year) & (hourly_avg_df['CONTAMINANT'] == contaminant)]

fig = px.line(filtered_df, x='Hour', y='Value', color='MUNICIPI',
              title=f'Hourly Average Concentrations of {contaminant} for Each City in {year}',
              labels={'Hour': 'Hour of the Day', 'Value': 'Average Concentration', 'MUNICIPI': 'City'})
fig.update_layout(legend_title_text='City')
fig.show()

fig.write_html('hourly_avg_no2_cities2024.html')


In [21]:
import plotly.express as px

avg_contamination_df = df.groupby(['NOM ESTACIO', 'LATITUD', 'LONGITUD', 'CONTAMINANT'])[['01h', '02h', '03h', '04h', '05h', '06h', '07h', '08h', '09h', '10h', '11h', '12h', '13h', '14h', '15h', '16h', '17h', '18h', '19h', '20h', '21h', '22h', '23h', '24h']].mean().reset_index()

avg_contamination_melted = avg_contamination_df.melt(id_vars=['NOM ESTACIO', 'LATITUD', 'LONGITUD', 'CONTAMINANT'], value_vars=[f'{i:02d}h' for i in range(1, 25)],
                                                     var_name='Hour', value_name='Average Value')

avg_contamination_melted = avg_contamination_melted[avg_contamination_melted['CONTAMINANT'] == contaminant]


fig = px.scatter_mapbox(avg_contamination_melted, lat='LATITUD', lon='LONGITUD', size='Average Value', color='Average Value',
                        hover_name='NOM ESTACIO', hover_data={'Average Value': True, 'Hour': True},
                        color_continuous_scale=px.colors.cyclical.IceFire, size_max=15, zoom=10,
                        title=f'Map Heatmap of {contaminant} Contamination Levels')

fig.update_layout(mapbox_style="open-street-map")
fig.update_layout(margin={"r":0,"t":0,"l":0,"b":0})
fig.show()


In [12]:
import pandas as pd
import plotly.graph_objects as go

# Sample
df_sampled = df.sample(100000, random_state=42)
df_sampled['DATA'] = pd.to_datetime(df_sampled['DATA'], format="%d/%m/%Y")

# Fixem any
fixed_year = 2024

# Filtrem dades per NO2
df_no2 = df_sampled[df_sampled['CONTAMINANT'] == 'NO2']

# Calcul dels promitjos diaris en un any
def get_daily_avg_data(year):
    filtered_df = df_no2[df_no2['DATA'].dt.year == year]
    filtered_df = filtered_df.melt(
        id_vars=['NOM ESTACIO', 'LATITUD', 'LONGITUD'],
        value_vars=[f'{i:02d}h' for i in range(1, 25)],
        var_name='Hour',
        value_name='Value'
    )
    filtered_df['Value'] = filtered_df['Value'].astype(float)
    daily_avg_df = filtered_df.groupby(['LATITUD', 'LONGITUD'])['Value'].mean().reset_index()
    return daily_avg_df

daily_avg_df = get_daily_avg_data(fixed_year)

# Plot mapa de densitat
fig = go.Figure(go.Densitymapbox(lat=daily_avg_df['LATITUD'], lon=daily_avg_df['LONGITUD'], z=daily_avg_df['Value'],
                                 radius=10, colorscale='Viridis', zmax=daily_avg_df['Value'].max(), zmin=daily_avg_df['Value'].min(),
                                 showscale=True))

fig.update_layout(mapbox_style="open-street-map", mapbox_center_lon=2.1734, mapbox_center_lat=41.3851,
                  mapbox_zoom=6, title=f'NO2 Pollution Density Map in {fixed_year}')

fig.update_layout(margin={"r":0,"t":30,"l":0,"b":0})
fig.show()

##fig.write_html('hourly_avg_no2_heatmap.html')


In [9]:
import pandas as pd
import plotly.express as px
import plotly.graph_objects as go

# Sample
df_sampled = df.sample(100000, random_state=42)

# Convertim
df_sampled['DATA'] = pd.to_datetime(df_sampled['DATA'], format="%d/%m/%Y")

#Filtrem Barcelona i NO2
df_barcelona_no2 = df_sampled[(df_sampled['MUNICIPI'] == 'Barcelona') & (df_sampled['CONTAMINANT'] == 'NO2')]

# Extraiem dades
df_barcelona_no2['Year'] = df_barcelona_no2['DATA'].dt.year
df_barcelona_no2['Hour'] = df_barcelona_no2['DATA'].dt.hour

# Agrupem i calculem nivells de NO2
hourly_avg_df = df_barcelona_no2.groupby(['Year', 'Hour'])[['01h', '02h', '03h', '04h', '05h', '06h',
                                                           '07h', '08h', '09h', '10h', '11h', '12h',
                                                           '13h', '14h', '15h', '16h', '17h', '18h',
                                                           '19h', '20h', '21h', '22h', '23h', '24h']].mean().reset_index()

hourly_avg_df = hourly_avg_df.melt(id_vars=['Year', 'Hour'], var_name='Hourly', value_name='Average Concentration')

# Definim escala de colors
color_scale = px.colors.sequential.Plasma
num_years = len(hourly_avg_df['Year'].unique())
colors = [color_scale[i * (len(color_scale) - 1) // (num_years - 1)] for i in range(num_years)]

fig = go.Figure()

# Afegim linies per cada any
for i, year in enumerate(hourly_avg_df['Year'].unique()):
    df_year = hourly_avg_df[hourly_avg_df['Year'] == year]
    fig.add_trace(go.Scatter(x=df_year['Hourly'], y=df_year['Average Concentration'],
                             mode='lines',
                             name=str(year),
                             line=dict(color=colors[i], width=2),
                             hovertemplate='Year: ' + str(year) + '<br>' +
                                           'Hour: %{x}<br>' +
                                           'Avg Concentration: %{y}<extra></extra>'
                             ))

# Afegim linies adicionals
fig.add_trace(go.Scatter(x=hourly_avg_df['Hourly'], y=[40] * len(hourly_avg_df),
                         mode='lines',
                         line=dict(color='red', width=1, dash='dash'),
                         name='European Limit for NO2 Concentration (40)',
                         hoverinfo='text',
                         hovertext='European Limit for NO2 Concentration: 40'
                         ))

fig.add_trace(go.Scatter(x=hourly_avg_df['Hourly'], y=[20] * len(hourly_avg_df),
                         mode='lines',
                         line=dict(color='blue', width=1, dash='dash'),
                         name='Objective for NO2 Concentration (20)',
                         hoverinfo='text',
                         hovertext='Objective for NO2 Concentration: 20'
                         ))

# Update layout
fig.update_layout(title='Hourly Average NO2 Concentrations Over Hours for Each Year in Barcelona',
                  xaxis_title='Hour of the Day',
                  yaxis_title='Average Concentration',
                  legend_title='Year')

fig.show()

##fig.write_html('hourly_avg_no2_barcelona.html')